pkg/proxy/iptables/proxier.go

```diff
@@ -1002,7 +998,82 @@ func (proxier *Proxier) syncProxyRules() {
 		// Service does not have conflicting configuration such as
 		// externalTrafficPolicy=Local.
 		allEndpoints = proxy.FilterEndpoints(allEndpoints, svcInfo, proxier.nodeLabels)
-		hasEndpoints := len(allEndpoints) > 0
+
+		// Scan the endpoints list to see what we have. "hasEndpoints" will be true
+		// if there are any usable endpoints for this service anywhere in the cluster.
+		var hasEndpoints, hasLocalReadyEndpoints, hasLocalServingTerminatingEndpoints bool
+		for _, ep := range allEndpoints {
+			if ep.IsReady() {
+				hasEndpoints = true
+				if ep.GetIsLocal() {
+					hasLocalReadyEndpoints = true
+				}
+			} else if svc.NodeLocalExternal() && utilfeature.DefaultFeatureGate.Enabled(features.ProxyTerminatingEndpoints) {
+				if ep.IsServing() && ep.IsTerminating() {
+					hasEndpoints = true
+					if ep.GetIsLocal() {
+						hasLocalServingTerminatingEndpoints = true
+					}
+				}
+			}
+		}
+		useTerminatingEndpoints := !hasLocalReadyEndpoints && hasLocalServingTerminatingEndpoints
+
+		// Generate the per-endpoint chains.
+		readyEndpointChains = readyEndpointChains[:0]
+		localEndpointChains = localEndpointChains[:0]
+		for _, ep := range allEndpoints {
+			epInfo, ok := ep.(*endpointsInfo)
+			if !ok {
+				klog.ErrorS(err, "Failed to cast endpointsInfo", "endpointsInfo", ep)
+				continue
+			}
```

Comment:
is this needed here? you iterate the same list in line 1005 without casting the endpoint

Reply:
Yes, we need to cast from `proxy.Endpoint` to `*endpointsInfo` in this loop so we can call `epInfo.endpointChain()`.

This is totally ugly and I'd like to get rid of it, but it's hard to do without either (a) moving `endpointChain()` into `proxy.Endpoint` even though it's iptables-specific, or (b) doing more memory allocations.

The `!ok` here is really a "can't happen" though, so in theory we could just make the cast blindly. But I didn't like doing that without adding checks somewhere to make sure that it _really_ can't happen. (It's not possible for a non-`*endpointsInfo` object to end up in that list, but it's not _obvious_ that it's not possible.)

Reply:
I think it's because we use the interface types acrss iptables and IPVS, and they have different impl types.  I don't like it either.